pkg/operator/mon/mon.go

```diff
@@ -36,187 +39,399 @@ const (
+			return fmt.Errorf("failed to save rook-admin secret. %+v", err)
 		}
 		logger.Infof("rook-admin secret already exists")
 	} else {
 		logger.Infof("saved rook-admin secret")
 	}
 
-	return info, nil
+	return nil
 }
 
-func (c *Cluster) startPods(clientset kubernetes.Interface, clusterInfo *mon.ClusterInfo, mons []*MonConfig) error {
+func (c *Cluster) startPods(conn client.Connection, mons []*MonConfig) error {
 	// schedule the mons on different nodes if we have enough nodes to be unique
-	antiAffinity, err := c.getAntiAffinity(clientset)
+	antiAffinity, err := c.getAntiAffinity()
 	if err != nil {
 		return fmt.Errorf("failed to get antiaffinity. %+v", err)
 	}
 
-	running, pending, err := c.pollPods(clientset, clusterInfo.Name)
-	if err != nil {
-		return fmt.Errorf("failed to get mon pods. %+v", err)
-	}
-	logger.Infof("%d running, %d pending pods", len(running), len(pending))
-
-	clusterInfo.Monitors = map[string]*mon.CephMonitorConfig{}
-	for _, m := range running {
-		clusterInfo.Monitors[m.Name] = mon.ToCephMon(m.Name, m.Status.PodIP)
-	}
-
-	if len(running) == c.Size {
-		logger.Infof("pods are already running")
-		return nil
-	}
-
-	started := 0
+	preexisted := len(c.clusterInfo.Monitors)
+	created := 0
 	alreadyRunning := 0
 	for _, m := range mons {
-		monPod := c.makeMonPod(m, clusterInfo, antiAffinity)
-		name := monPod.Name
+		monPod := c.makeMonPod(m, antiAffinity)
 		logger.Debugf("Starting pod: %+v", monPod)
-		monPod, err = clientset.CoreV1().Pods(c.Namespace).Create(monPod)
+		name := monPod.Name
+		_, err = c.clientset.CoreV1().Pods(c.Namespace).Create(monPod)
 		if err != nil {
 			if !errors.IsAlreadyExists(err) {
 				return fmt.Errorf("failed to create mon pod %s. %+v", name, err)
 			}
+			logger.Infof("pod %s already exists", name)
 			alreadyRunning++
 		} else {
-			started++
+			created++
 		}
 
-		podIP, err := c.waitForPodToStart(clientset, name)
+		podIP, err := c.waitForPodToStart(name)
 		if err != nil {
 			return fmt.Errorf("failed to start pod %s. %+v", name, err)
 		}
-		clusterInfo.Monitors[m.Name] = mon.ToCephMon(m.Name, podIP)
+		c.clusterInfo.Monitors[m.Name] = mon.ToCephMon(m.Name, podIP)
+
+		err = c.saveMonConfig()
+		if err != nil {
+			return fmt.Errorf("failed to save endpoints after starting mon %s. %+v", m.Name, err)
+		}
+	}
+
+	logger.Infof("mons created: %d, alreadyRunning: %d, preexisted: %d", created, alreadyRunning, preexisted)
+
+	return c.waitForMonsToJoin(conn, mons)
+}
+
+func (c *Cluster) waitForMonsToJoin(conn client.Connection, mons []*MonConfig) error {
+	if !c.waitForStart {
+		return nil
+	}
+
+	// initialize a connection if it is not already connected
+	if conn == nil {
+		context := &clusterd.Context{ConfigDir: k8sutil.DataDir}
+		var err error
+		conn, err = mon.ConnectToClusterAsAdmin(context, c.factory, c.clusterInfo)
+		if err != nil {
+			return fmt.Errorf("cannot connect to cluster. %+v", err)
+		}
+		defer conn.Shutdown()
```

Comment:
is this defer correct here?  the caller has passed in a connection ref and if it's not created yet we create it here.  does the caller expect to use the connection after this func returns? if so, we do not want to do this defer here as the caller will have an already shutdown connection.